spire-agent panic when using prometheus telemetry

After enabling prometheus telemetry, spire-agent panics reliably when the liveness probe calls `api fetch`:

```
panic: counter cannot decrease in value

goroutine 199 [running]:
github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc0005d3ac0, 0xbff0000000000000)
    /go/pkg/mod/github.com/prometheus/client_golang@v0.9.3-0.20190127221311-3c4408c8b829/prometheus/counter.go:89 +0x134
github.com/armon/go-metrics/prometheus.(*PrometheusSink).IncrCounterWithLabels(0xc000263770, 0xc0007aa0c0, 0x3, 0x4, 0xc0bf800000, 0xc00077e500, 0x1, 0x1)
    /go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20180917152333-f0300d1749da/prometheus/prometheus.go:192 +0x15b
github.com/armon/go-metrics.FanoutSink.IncrCounterWithLabels(0xc0000bd780, 0x2, 0x2, 0xc0007aa0c0, 0x3, 0x4, 0xc0bf800000, 0xc00077e500, 0x1, 0x1)
    /go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20180917152333-f0300d1749da/sink.go:63 +0xbc
github.com/armon/go-metrics.(*Metrics).IncrCounterWithLabels(0xc00002e0f0, 0xc00077e4e0, 0x3, 0x4, 0xc0bf800000, 0x0, 0x1, 0x1)
    /go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20180917152333-f0300d1749da/metrics.go:81 +0x305
github.com/armon/go-metrics.(*Metrics).IncrCounter(0xc00002e0f0, 0xc00077e4e0, 0x2, 0x2, 0xc0bf800000)
    /go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20180917152333-f0300d1749da/metrics.go:60 +0x66
github.com/spiffe/spire/pkg/agent/endpoints/workload.(*Handler).startCall.func1()
    /spire/pkg/agent/endpoints/workload/handler.go:363 +0xa7
github.com/spiffe/spire/pkg/agent/endpoints/workload.(*Handler).FetchX509SVID(0xc0001e93c0, 0xc000710860, 0x1cc2da0, 0xc0005c3ba0, 0x1c6b6c0, 0xc0007a0230)
    /spire/pkg/agent/endpoints/workload/handler.go:206 +0x546
github.com/spiffe/spire/proto/spire/api/workload._SpiffeWorkloadAPI_FetchX509SVID_Handler(0x18c4b60, 0xc0001e93c0, 0x1cbb660, 0xc00033e9c0, 0x2b5cfb8, 0xc0002f9600)
    /spire/proto/spire/api/workload/workload.pb.go:753 +0x109
google.golang.org/grpc.(*Server).processStreamingRPC(0xc00049f680, 0x1ccdd80, 0xc00071e780, 0xc0002f9600, 0xc000196d20, 0x2b29760, 0x0, 0x0, 0x0)
    /go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:1175 +0xacd
google.golang.org/grpc.(*Server).handleStream(0xc00049f680, 0x1ccdd80, 0xc00071e780, 0xc0002f9600, 0x0)
    /go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:1254 +0xcbe
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0009fdac0, 0xc00049f680, 0x1ccdd80, 0xc00071e780, 0xc0002f9600)
    /go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:690 +0x9f
created by google.golang.org/grpc.(*Server).serveStreams.func1
    /go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:688 +0xa1
```

The preceding log entry (from the `api-fetch` call is):

```
time="2019-05-30T22:44:25Z" level=debug msg="PID 29406 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\"  type:\"k8s\" value:\"sa:spire-agent\"  type:\"k8s\" value:\"ns:spire\"  type:\"k8s\" value:\"node-name:ip-172-20-32-204.us-east-2.compute.internal\"  type:\"k8s\" value:\"pod-uid:ecce151c-832a-11e9-b496-02d84beb7d12\"  type:\"k8s\" value:\"container-name:spire-agent\"  type:\"k8s\" value:\"container-image:gcr.io/spiffe-io/spire-agent:0.8.0\"  type:\"k8s\" value:\"pod-label:app:spire-agent\"  type:\"k8s\" value:\"pod-label:controller-revision-hash:668dd5946d\"  type:\"k8s\" value:\"pod-label:pod-template-generation:1\"  type:\"k8s\" value:\"pod-owner:DaemonSet:spire-agent\"  type:\"k8s\" value:\"pod-owner-uid:DaemonSet:ecc9eb8e-832a-11e9-b496-02d84beb7d12\" ]" subsystem_name=workload_api
```

This was observed using the `gcr.io/spiffe-io/spire-agent:0.8.0` container image.

--- a/pkg/agent/endpoints/workload/handler.go
+++ b/pkg/agent/endpoints/workload/handler.go
@@ -9,6 +9,7 @@ import (
 	"errors"
 	"fmt"
 	"strings"
+	"sync/atomic"
 	"time"
 
 	"github.com/golang/protobuf/jsonpb"
@@ -41,6 +42,9 @@ type Handler struct {
 	Catalog catalog.Catalog
 	Log     logrus.FieldLogger
 	Metrics telemetry.Metrics
+
+	// tracks the number of outstanding connections
+	connections int32
 }
 
 // FetchJWTSVID processes request for a JWT SVID
@@ -322,7 +326,11 @@ func (h *Handler) startCall(ctx context.Context) (int32, []*common.Selector, tel
 	}
 
 	// add to count of current
-	telemetry_workload.IncrConnectionTotalCounter(h.Metrics)
+	telemetry_workload.SetConnectionTotalGauge(h.Metrics, atomic.AddInt32(&h.connections, 1))
+	done := func() {
+		// rely on caller to decrement count of current connections
+		telemetry_workload.SetConnectionTotalGauge(h.Metrics, atomic.AddInt32(&h.connections, -1))
+	}
 
 	config := attestor.Config{
 		Catalog: h.Catalog,
@@ -334,20 +342,13 @@ func (h *Handler) startCall(ctx context.Context) (int32, []*common.Selector, tel
 
 	// Ensure that the original caller is still alive so that we know we didn't
 	// attest some other process that happened to be assigned the original PID
-	err = watcher.IsAlive()
-	if err != nil {
-		// error, decrement count of current connections
-		telemetry_workload.DecrConnectionTotalCounter(h.Metrics)
+	if err := watcher.IsAlive(); err != nil {
+		done()
 		return 0, nil, nil, nil, status.Errorf(codes.Unauthenticated, "Could not verify existence of the original caller: %v", err)
 	}
 
 	metrics := telemetry.WithLabels(h.Metrics, selectorsToLabels(selectors))
 	telemetry_workload.IncrConnectionCounter(metrics)
-
-	done := func() {
-		// rely on caller to decrement count of current connections
-		telemetry_workload.DecrConnectionTotalCounter(h.Metrics)
-	}
 
 	return watcher.PID(), selectors, metrics, done, nil
 }

--- a/pkg/agent/endpoints/workload/handler_test.go
+++ b/pkg/agent/endpoints/workload/handler_test.go
@@ -360,8 +360,8 @@ func setupMetricsCommonExpectations(metrics *mock_telemetry.MockMetrics, selecto
 	metrics.EXPECT().MeasureSince([]string{telemetry.WorkloadAPI, telemetry.WorkloadAttestationDuration}, gomock.Any())
 
 	metrics.EXPECT().IncrCounterWithLabels([]string{telemetry.WorkloadAPI, telemetry.Connection}, float32(1), selectorsLabels)
-	metrics.EXPECT().IncrCounter([]string{telemetry.WorkloadAPI, telemetry.Connections}, float32(1))
-	metrics.EXPECT().IncrCounter([]string{telemetry.WorkloadAPI, telemetry.Connections}, float32(-1))
+	metrics.EXPECT().SetGauge([]string{telemetry.WorkloadAPI, telemetry.Connections}, float32(1))
+	metrics.EXPECT().SetGauge([]string{telemetry.WorkloadAPI, telemetry.Connections}, float32(0))
 }
 
 func (s *HandlerTestSuite) TestFetchJWTBundles() {

--- a/pkg/common/telemetry/agent/workloadapi/workload.go
+++ b/pkg/common/telemetry/agent/workloadapi/workload.go
@@ -43,16 +43,9 @@ func IncrConnectionCounter(m telemetry.Metrics) {
 	m.IncrCounter([]string{telemetry.WorkloadAPI, telemetry.Connection}, 1)
 }
 
-// DecrConnectionTotalCounter indicate one less
-// Workload API active connection (active in that moment)
-func DecrConnectionTotalCounter(m telemetry.Metrics) {
-	m.IncrCounter([]string{telemetry.WorkloadAPI, telemetry.Connections}, -1)
-}
-
-// IncrConnectionTotalCounter indicate one more
-// Workload API active connection (active in that moment)
-func IncrConnectionTotalCounter(m telemetry.Metrics) {
-	m.IncrCounter([]string{telemetry.WorkloadAPI, telemetry.Connections}, 1)
+// SetConnectionTotalGauge sets the number of active Workload API connections
+func SetConnectionTotalGauge(m telemetry.Metrics, connections int32) {
+	m.SetGauge([]string{telemetry.WorkloadAPI, telemetry.Connections}, float32(connections))
 }
 
 // IncrFetchJWTBundlesCounter indicate call to Workload

